File: 4_data_structures_and_libraries/2893.py
# https://www.urionlinejudge.com.br/judge/pt/problems/view/2893

T = int(input())

while T:
    k, N = (int(x) for x in input().split(" "))
    
    fibo = [0] + [None] * 20000
    summ = [0] + [None] * 20000
    
    for i in range(1, k + 1):
        fibo[i] = i
        summ[i] = summ[i - 1] + i
    
    j = 0
    for i in range(k, N):
        fibo[i] = summ[i - 1]
        summ[i] = summ[i - 1] + fibo[i] - fibo[j]
        j += 1

    print(fibo[N - 1] % 1000007)

    T -= 1

File: 4_data_structures_and_libraries/2893.c
// https://www.urionlinejudge.com.br/judge/pt/problems/view/2893

#include <stdio.h>

int mod(int x, int n)
{
    return (x % n + n) % n;
}

int main()
{
    int k, N, T, i, j;
    int fib[20000];
    int sum[20000];

    scanf("%d", &T);

    while (T) {
        scanf("%d%d", &k, &N);

        fib[0] = sum[0] = 0;
        for (i = 1; i < k; ++i) {
            fib[i] = i;
            sum[i] = sum[i - 1] + i;
        }

        for (i = k, j = 0; i < N; ++i, ++j) {
            fib[i] = sum[i - 1];
            sum[i] = mod(mod(sum[i - 1] + fib[i], 1000007) - fib[j], 1000007);
        }

        printf("%d\n", fib[N - 1]);

        --T;
    }

    return 0;
}
